File: basico/volume/cone/02-upgrade/volume_cone.py
"""
Animação do Cálculo do Volume de um Cone - Método das Cascas Cilíndricas
=========================================================================

Mostra visualmente como calcular V = (1/3)πr²h usando integração por
cascas cilíndricas variando ρ de 0 até r.

Autor: Ivan Carlos Lopes
"""

from matplotlib import pyplot, animation, cm, patches, path
import numpy as np

# Configuração de fontes para renderização matemática elegante
pyplot.rcParams.update({"font.family": "sans-serif", "mathtext.fontset": "cm"})


def c32(xyz):
    """
    Projeção isométrica (cavaleira) 3D → 2D

    Vetorizada para aceitar arrays numpy
    """
    return np.array([
        xyz[1] - xyz[0] * np.sin(np.pi/4) / 2,
        xyz[2] - xyz[0] * np.cos(np.pi/4) / 2
    ])


# ==============================================================================
# CONFIGURAÇÃO DA ANIMAÇÃO
# ==============================================================================

Nf = 2900  # Número de frames
cl = list(cm.tab10.colors)

# Criação da figura
fig = pyplot.figure(figsize=(9, 9))

# Limites de visualização (zoom)
xlim = [-0.9, 0.9]
ylim = [-0.35, 1.15]

# Dois eixos sobrepostos
ax1 = fig.add_axes([0, 0, 1, 1], xlim=xlim, ylim=ylim, fc='none')
ax1.axis('off')

ax = fig.add_axes([0, 0, 1, 1], xlim=xlim, ylim=ylim, fc='none')
ax.axis('off')

# Título
ax.text(0, ylim[1] - 0.05, 'Método das Cascas Cilíndricas',
        size=35, ha='center', va='top', weight='bold')


# ==============================================================================
# PARÂMETROS GEOMÉTRICOS
# ==============================================================================

# Dimensões do cone
r = 0.5   # raio da base
h = 0.8   # altura

# ==============================================================
# POSIÇÃO GLOBAL DO DESENHO (AJUSTE ANTES DE RODAR)
# ==============================================================

X_POS = 0.0    # deslocamento horizontal
Y_POS = 0.05    # deslocamento vertical

# Posição central
xy = np.array([X_POS, Y_POS])

# Tamanho dos eixos
xm, ym, zm = 0.8, 0.8, 0.9

# Ângulos de visibilidade
t1 = np.pi - np.atan(2*np.sqrt(2))
t2 = t1 + np.pi
theta = np.linspace(t1, t2, 100)

# ⚡ PRÉ-CALCULA valores que não mudam
theta_plus_pi = theta + np.pi
cos_theta = np.cos(theta)
sin_theta = np.sin(theta)
cos_theta_pi = np.cos(theta_plus_pi)
sin_theta_pi = np.sin(theta_plus_pi)

# Array de raios para animação
rho = np.linspace(0, r, Nf)

# ⚡ PRÉ-CALCULA coordenadas da base do cone (não muda)
base_visible_x = xy[0] + r * \
    c32([cos_theta, sin_theta, np.zeros_like(theta)])[0]
base_visible_y = xy[1] + r * \
    c32([cos_theta, sin_theta, np.zeros_like(theta)])[1]

base_hidden_x = xy[0] + r * \
    c32([-cos_theta, -sin_theta, np.zeros_like(theta)])[0]
base_hidden_y = xy[1] + r * \
    c32([-cos_theta, -sin_theta, np.zeros_like(theta)])[1]

# ⚡ PRÉ-CALCULA arestas laterais
edge_points = []
for i in range(2):
    angle = t1 + np.pi * i
    base_pt = xy + c32([r*np.cos(angle), r*np.sin(angle), 0])
    apex_pt = xy + c32([0, 0, h])
    edge_points.append((base_pt, apex_pt))


# ==============================================================================
# DESENHO DA ESTRUTURA ESTÁTICA
# ==============================================================================

# Eixos de coordenadas
axes = [c32([xm, 0, 0]), c32([0, ym, 0]), c32([0, 0, zm])]
for axis_vec in axes:
    ax.annotate('', xy + axis_vec, xy,
                arrowprops=dict(arrowstyle='->', lw=1, color='k'))

# Contorno da base (pré-calculado)
ax.plot(base_visible_x, base_visible_y, '--', c=cl[0], lw=1.2, alpha=.6)
ax.plot(base_hidden_x, base_hidden_y, '-', c=cl[0], lw=2, alpha=.6)

# Arestas laterais (pré-calculadas)
for base_pt, apex_pt in edge_points:
    ax.plot([base_pt[0], apex_pt[0]], [base_pt[1], apex_pt[1]],
            c=cl[0], lw=1, alpha=.6)

# Linha do raio com label
r_end = xy + c32([r*np.cos(t1), r*np.sin(t1), 0])
ax.plot([xy[0], r_end[0]], [xy[1], r_end[1]], '--', c='k', lw=1.2)
ax.text((xy[0] + r_end[0])/2, (xy[1] + r_end[1])/2,
        '$r$', size=25, c='k', ha='center', va='bottom')

# Label de altura
ax.text(xy[0], xy[1] + h/2, '$h$', size=25, c='k', ha='right', va='center')

# Fórmulas matemáticas
formulas_y = ylim[0] + 0.1

# ax.text(0, formulas_y,
#         r'$dV = 2\pi\rho \cdot h\left(1-\frac{\rho}{r}\right) d\rho$',
#         size=20, ha='center', va='bottom',
#         bbox=dict(boxstyle='round', pad=0.4, facecolor='lightyellow',
#                   edgecolor=cl[1], lw=2.5))

# ax.text(0, formulas_y - 0.15,
#         r'$V = \int_0^r 2\pi\rho\left(h - \frac{h}{r}\rho\right) d\rho = \frac{1}{3}\pi r^2 h$',
#         size=22, ha='center', va='bottom', weight='bold',
#         bbox=dict(boxstyle='round', pad=0.5, facecolor='lightgreen',
#                   edgecolor=cl[2], lw=3))


# ==============================================================================
# FUNÇÃO DE ANIMAÇÃO OTIMIZADA
# ==============================================================================

# ⚡ Objetos que serão atualizados (evita recriar em cada frame)
shell_patch = None
top_circle_line = None
volume_patch = None

# ⚡ PRÉ-CALCULA códigos de path (não mudam)
n_theta = len(theta_plus_pi)
# codes_shell: base (n_theta) + topo (n_theta) + fechamento (1) = 2*n_theta + 1
codes_shell = [path.Path.MOVETO] + [path.Path.LINETO] * \
    (2*n_theta - 1) + [path.Path.CLOSEPOLY]
# codes_volume: círculo (n_theta) + apex (1) + fechamento (1) = n_theta + 2
codes_volume = [path.Path.MOVETO] + \
    [path.Path.LINETO]*(n_theta) + [path.Path.CLOSEPOLY]


def animate(i):
    """Animação otimizada - minimiza recriação de objetos"""
    global shell_patch, top_circle_line, volume_patch

    if i % 25 == 0:
        print(f"Frame {i}/{Nf}")

    # Limpa apenas os patches dinâmicos
    ax1.cla()
    ax1.set_xlim(xlim)
    ax1.set_ylim(ylim)
    ax1.axis('off')

    rho_i = rho[i]

    # ⚡ Evita divisão por zero
    if rho_i < 1e-6:
        return []

    z_top = h * (1 - rho_i / r)

    # ===========================================================================
    # CASCA CILÍNDRICA (vetorizado)
    # ===========================================================================

    # ⚡ Calcula vértices de forma vetorizada
    verts = np.zeros((2*n_theta + 1, 2))

    # Base (círculo inferior)
    base_coords = c32([rho_i * cos_theta_pi, rho_i * sin_theta_pi,
                       np.zeros(n_theta)])
    verts[:n_theta, 0] = xy[0] + base_coords[0]
    verts[:n_theta, 1] = xy[1] + base_coords[1]

    # Topo (círculo superior) - ordem reversa
    top_coords = c32([rho_i * cos_theta_pi[::-1],
                     rho_i * sin_theta_pi[::-1],
                     np.full(n_theta, z_top)])
    verts[n_theta:2*n_theta, 0] = xy[0] + top_coords[0]
    verts[n_theta:2*n_theta, 1] = xy[1] + top_coords[1]

    # Fecha o polígono
    verts[-1] = [0, 0]

    # Adiciona patch da casca
    ax1.add_patch(patches.PathPatch(
        path.Path(verts, codes_shell),
        fc=cl[1], ec='none', alpha=0.8, zorder=1
    ))

    # ===========================================================================
    # CÍRCULO SUPERIOR (tracejado)
    # ===========================================================================

    top_circle_coords = c32([rho_i * cos_theta, rho_i * sin_theta,
                             np.full(len(theta), z_top)])

    ax1.plot(xy[0] + top_circle_coords[0],
             xy[1] + top_circle_coords[1],
             '--', c=cl[1], lw=1.2, zorder=-2)

    # ===========================================================================
    # VOLUME ACUMULADO (vetorizado)
    # ===========================================================================

    verts_vol = np.zeros((n_theta + 2, 2))

    # Círculo no topo da casca
    verts_vol[:n_theta, 0] = xy[0] + top_coords[0][::-1]  # reverte de volta
    verts_vol[:n_theta, 1] = xy[1] + top_coords[1][::-1]

    # Vértice do cone
    apex = xy + c32([0, 0, h])
    verts_vol[n_theta] = apex

    # Fecha o polígono
    verts_vol[-1] = [0, 0]

    ax1.add_patch(patches.PathPatch(
        path.Path(verts_vol, codes_volume),
        fc=cl[0], ec='none', alpha=0.8, zorder=1
    ))

    return []


# ==============================================================================
# RENDERIZAÇÃO E SALVAMENTO
# ==============================================================================

# Assinatura
ax.text(0, ylim[0] + 0.01,
        r'@Ivan Carlos Lopes',
        size=12, c='.2', alpha=.3, ha='center', va='bottom',
        style='italic')

print("Criando animação...")
anim = animation.FuncAnimation(
    fig, animate, frames=Nf, interval=20, blit=False)

print("Salvando animação (pode demorar alguns minutos)...")
anim.save("volume_cone_optimized.mp4",
          writer=animation.FFMpegWriter(fps=60),
          dpi=200)
print("✓ Animação salva: volume_cone_optimized.mp4")

File: basico/volume/cone/cpp/21/volume_cone.cpp
#include "matplotlibcpp.h"
#include <cmath>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <vector>

namespace plt = matplotlibcpp;

const double PI = 3.14159265358979323846;
const int    Nf = 100;  // Número de frames
const double r  = 0.5;  // raio da base
const double h  = 0.8;  // altura

struct Point2D {
    double x, y;
};

// Projeção isométrica 3D -> 2D
Point2D c32(double x3d, double y3d, double z3d) {
    return {y3d - x3d * std::sin(PI / 4.0) / 2.0,
            z3d - x3d * std::cos(PI / 4.0) / 2.0};
}

std::vector<double> linspace(double start, double end, int num) {
    std::vector<double> result(num);
    double step = (end - start) / (num - 1);
    for(int i = 0; i < num; ++i) {
        result[i] = start + i * step;
    }
    return result;
}

struct ConeSetup {
    double t1, t2;
    std::vector<double> theta;
    std::vector<double> theta_plus_pi;
    std::vector<double> cos_theta;
    std::vector<double> sin_theta;
    std::vector<double> cos_theta_pi;
    std::vector<double> sin_theta_pi;
    std::vector<double> rho;
    Point2D xy;

    ConeSetup() : xy{0.0, 0.05} {
        t1 = PI - std::atan(2.0 * std::sqrt(2.0));
        t2 = t1 + PI;
        theta = linspace(t1, t2, 100);

        theta_plus_pi.resize(theta.size());
        cos_theta.resize(theta.size());
        sin_theta.resize(theta.size());
        cos_theta_pi.resize(theta.size());
        sin_theta_pi.resize(theta.size());

        for(size_t i = 0; i < theta.size(); ++i) {
            theta_plus_pi[i] = theta[i] + PI;
            cos_theta[i] = std::cos(theta[i]);
            sin_theta[i] = std::sin(theta[i]);
            cos_theta_pi[i] = std::cos(theta_plus_pi[i]);
            sin_theta_pi[i] = std::sin(theta_plus_pi[i]);
        }

        rho = linspace(0, r, Nf);
    }
};

void drawStaticStructure(const ConeSetup& setup) {
    // Contorno da base (parte visível - tracejada)
    std::vector<double> base_vis_x, base_vis_y;
    for(size_t i = 0; i < setup.theta.size(); ++i) {
        Point2D pt = c32(r * setup.cos_theta[i], r * setup.sin_theta[i], 0);
        base_vis_x.push_back(setup.xy.x + pt.x);
        base_vis_y.push_back(setup.xy.y + pt.y);
    }

    std::map<std::string, std::string> kw_vis;
    kw_vis["linestyle"] = "--";
    kw_vis["color"] = "blue";
    kw_vis["linewidth"] = "1.2";
    kw_vis["alpha"] = "0.6";
    plt::plot(base_vis_x, base_vis_y, kw_vis);

    // Contorno da base (parte oculta - contínua)
    std::vector<double> base_hid_x, base_hid_y;
    for(size_t i = 0; i < setup.theta.size(); ++i) {
        Point2D pt = c32(-r * setup.cos_theta[i], -r * setup.sin_theta[i], 0);
        base_hid_x.push_back(setup.xy.x + pt.x);
        base_hid_y.push_back(setup.xy.y + pt.y);
    }

    std::map<std::string, std::string> kw_hid;
    kw_hid["color"] = "blue";
    kw_hid["linewidth"] = "2";
    kw_hid["alpha"] = "0.6";
    plt::plot(base_hid_x, base_hid_y, kw_hid);

    // Arestas laterais
    for(int i = 0; i < 2; ++i) {
        double angle = setup.t1 + PI * i;
        Point2D base_pt = c32(r * std::cos(angle), r * std::sin(angle), 0);
        Point2D apex_pt = c32(0, 0, h);

        std::vector<double> edge_x = {setup.xy.x + base_pt.x, setup.xy.x + apex_pt.x};
        std::vector<double> edge_y = {setup.xy.y + base_pt.y, setup.xy.y + apex_pt.y};

        std::map<std::string, std::string> kw_edge;
        kw_edge["color"] = "blue";
        kw_edge["linewidth"] = "1";
        kw_edge["alpha"] = "0.6";
        plt::plot(edge_x, edge_y, kw_edge);
    }
}

void drawShell(const ConeSetup& setup, int i) {
    double rho_i = setup.rho[i];
    if(rho_i < 1e-6) return;

    double z_top = h * (1.0 - rho_i / r);

    // Casca cilíndrica
    std::vector<double> shell_x, shell_y;

    // Base (círculo inferior)
    for(size_t j = 0; j < setup.theta_plus_pi.size(); ++j) {
        Point2D pt = c32(rho_i * setup.cos_theta_pi[j],
                         rho_i * setup.sin_theta_pi[j], 0);
        shell_x.push_back(setup.xy.x + pt.x);
        shell_y.push_back(setup.xy.y + pt.y);
    }

    // Topo (círculo superior) - ordem reversa
    for(int j = setup.theta_plus_pi.size() - 1; j >= 0; --j) {
        Point2D pt = c32(rho_i * setup.cos_theta_pi[j],
                         rho_i * setup.sin_theta_pi[j], z_top);
        shell_x.push_back(setup.xy.x + pt.x);
        shell_y.push_back(setup.xy.y + pt.y);
    }

    shell_x.push_back(shell_x[0]);
    shell_y.push_back(shell_y[0]);

    // Desenha casca preenchida
    std::map<std::string, std::string> kw_shell;
    kw_shell["color"] = "orange";
    kw_shell["alpha"] = "0.6";
    plt::fill(shell_x, shell_y, kw_shell);

    // Círculo superior (tracejado)
    std::vector<double> top_x, top_y;
    for(size_t j = 0; j < setup.theta.size(); ++j) {
        Point2D pt = c32(rho_i * setup.cos_theta[j],
                         rho_i * setup.sin_theta[j], z_top);
        top_x.push_back(setup.xy.x + pt.x);
        top_y.push_back(setup.xy.y + pt.y);
    }

    std::map<std::string, std::string> kw_top;
    kw_top["linestyle"] = "--";
    kw_top["color"] = "orange";
    kw_top["linewidth"] = "1.2";
    plt::plot(top_x, top_y, kw_top);

    // Volume acumulado (cone interno)
    std::vector<double> vol_x, vol_y;

    for(size_t j = 0; j < setup.theta_plus_pi.size(); ++j) {
        Point2D pt = c32(rho_i * setup.cos_theta_pi[j],
                         rho_i * setup.sin_theta_pi[j], z_top);
        vol_x.push_back(setup.xy.x + pt.x);
        vol_y.push_back(setup.xy.y + pt.y);
    }

    Point2D apex = c32(0, 0, h);
    vol_x.push_back(setup.xy.x + apex.x);
    vol_y.push_back(setup.xy.y + apex.y);
    vol_x.push_back(vol_x[0]);
    vol_y.push_back(vol_y[0]);

    std::map<std::string, std::string> kw_vol;
    kw_vol["color"] = "lightblue";
    kw_vol["alpha"] = "0.7";
    plt::fill(vol_x, vol_y, kw_vol);
}

int main() {
    std::cout << "Gerando frames do cone..." << std::endl;

    // Configura backend não-interativo
    plt::backend("Agg");

    ConeSetup setup;
    system("mkdir -p frames");

    for(int i = 0; i < Nf; ++i) {
        if(i % 10 == 0) {
            std::cout << "Frame " << i << "/" << Nf << std::endl;
        }

        plt::clf();
        plt::xlim(-0.9, 0.9);
        plt::ylim(-0.35, 1.15);
        plt::axis("off");

        drawStaticStructure(setup);
        drawShell(setup, i);

        std::ostringstream filename;
        filename << "frames/frame_" << std::setfill('0')
                 << std::setw(4) << i << ".png";
        plt::save(filename.str());
    }

    std::cout << "\n✓ Frames gerados em ./frames/" << std::endl;
    std::cout << "\nPara criar video:" << std::endl;
    std::cout << "ffmpeg -framerate 30 -i frames/frame_%04d.png "
                 "-c:v libx264 -pix_fmt yuv420p cone.mp4" << std::endl;

    return 0;
}
